# UCL AI Society Machine Learning Tutorials
### Session 01. Introduction to Numpy, Pandas, Matplotlib

### Contents
1. Numpy
2. Pandas
3. Matplotlib
4. EDA(Exploratory Data Analysis)

### Aim
At the end of this session, you will be able to:
- Understand the basics of numpy.
- Understand the basics of pandas.
- Understand the basics of matplotlib.
- Perform an Exploratory Data Analysis (EDA).


## 2. Pandas
Pandas is another essential open-source library of Python and today it is widely used by data scientists and analysts. It is built by Wes McKinney based on numpy. The name 'Pandas' is originated from the term "Panel Data", an econometrics term for data sets that include observations over multiple time periods for the same individual.

### 2.1 Basics of Pandas

In [1]:
# run this shell if you haven't install pandas library
! pip install pandas

You should consider upgrading via the 'pip install --upgrade pip' command.


In [2]:
import pandas as pd
import numpy as np

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/matplotlib/__init__.py:1003: UserWarning: Duplicate key in file "/Users/toeun_kim/.matplotlib/matplotlibrc", line #2
  (fname, cnt))


In [3]:
print(pd.__version__)

0.24.2


The main data structures of pandas are **Series** and **DataFrames**, where data are stored and manipulated. A `Series` is a column and a `DataFrame` is a multi-dimensional table that consists of a collection of Series.

In [4]:
a = pd.Series([1, 2, 3, np.nan, 5, 6])

In [5]:
print(a)

0    1.0
1    2.0
2    3.0
3    NaN
4    5.0
5    6.0
dtype: float64


**Let's see how they are different!**

In [22]:
# TODO: Make Series using pandas Series()
# This is another way to initialise a pandas series
module_score_dic = {'Database': 90, 'Security': 70, 'Math': 100, 'Machine Learning': 80}
module_score = pd.Series(module_score_dic)
print("Module_score: \n", module_score, '\n')
print("type: \n", type(module_score), '\n')

# TODO: Make DataFrame using pandas DataFrame()
dataframe = pd.DataFrame(module_score, columns=['score'])
# dataframe = pd.DataFrame(module_score, index=[x for x in module_score.keys()], columns=['score'])
print("dataframe: \n", dataframe, '\n')
print("type: \n", type(data))
dataframe.dtypes

Module_score: 
 Database             90
Security             70
Math                100
Machine Learning     80
dtype: int64 

type: 
 <class 'pandas.core.series.Series'> 

dataframe: 
                   score
Database             90
Security             70
Math                100
Machine Learning     80 

type: 
 <class 'pandas.core.frame.DataFrame'>


score    int64
dtype: object

Series can also be a Dataframe that has only one attribute.  
**Now let's make a Dataframe that has multiple attributes**

In [18]:
solar_data = {
    'Name' : ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune"],
    'Satellite' : [0, 0, 1, 2, 79, 60, 27, 14],
    'AU' : [0.4, 0.7, 1, 1.5, 5.2, 9.5, 19.2, 30.1],
    'Diameter (in 1Kkm)' : [4.9, 12.1, 12.7, 6.8, 139.8, 116.5, 50.7, 49.2]
}

In [19]:
solar_system = pd.DataFrame(solar_data, index = [i for i in range(1, 9)])
solar_system

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1
3    Earth          1   1.0                12.7
4     Mars          2   1.5                 6.8
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [23]:
solar_system.dtypes # check data type

Name                   object
Satellite               int64
AU                    float64
Diameter (in 1Kkm)    float64
dtype: object

We can select what to read from the DataFrame
- `head()` : Extracts the first few data
- `tail()` : Extracts the last few data
- `index` : Extracts the index
- `columns` : Extracts the column
- `loc` : Extracts the information of that row
- `values` : Extracts only the values
- `describe()` : Outputs the summary of statistics of DataFrame
- `sort_values(self, by, axis = 0, ascending = True, inplace = False)` : Sort the DataFrame
- `drop()` : Drop the selected row

In [24]:
df = solar_system

In [25]:
df.head() # the default value in the bracket is 5

Name  Satellite   AU  Diameter (in 1Kkm)
1  Mercury          0  0.4                 4.9
2    Venus          0  0.7                12.1
3    Earth          1  1.0                12.7
4     Mars          2  1.5                 6.8
5  Jupiter         79  5.2               139.8

In [26]:
df.tail(2)

Name  Satellite    AU  Diameter (in 1Kkm)
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [40]:
df.index

Int64Index([1, 2, 3, 4, 5, 6, 7, 8, 9], dtype='int64')

**Beware! 0th index of df.index is 1**

In [39]:
df.index[0]

1

In [28]:
df.columns

Index(['Name', 'Satellite', 'AU', 'Diameter (in 1Kkm)'], dtype='object')

In [30]:
# df.loc[0] gives you an error. number inside bracket doesn't work like an indexing in list. you have to put df.index
df.loc[3]

Name                  Earth
Satellite                 1
AU                        1
Diameter (in 1Kkm)     12.7
Name: 3, dtype: object

In [31]:
df.values

array([['Mercury', 0, 0.4, 4.9],
       ['Venus', 0, 0.7, 12.1],
       ['Earth', 1, 1.0, 12.7],
       ['Mars', 2, 1.5, 6.8],
       ['Jupiter', 79, 5.2, 139.8],
       ['Saturn', 60, 9.5, 116.5],
       ['Uranus', 27, 19.2, 50.7],
       ['Neptune', 14, 30.1, 49.2]], dtype=object)

In [32]:
df.describe()

Satellite         AU  Diameter (in 1Kkm)
count   8.000000   8.000000             8.00000
mean   22.875000   8.450000            49.08750
std    30.670775  10.853702            52.38417
min     0.000000   0.400000             4.90000
25%     0.750000   0.925000            10.77500
50%     8.000000   3.350000            30.95000
75%    35.250000  11.925000            67.15000
max    79.000000  30.100000           139.80000

In [35]:
df.sort_values(by = 'Diameter (in 1Kkm)', ascending = False)

Name  Satellite    AU  Diameter (in 1Kkm)
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2
3    Earth          1   1.0                12.7
2    Venus          0   0.7                12.1
4     Mars          2   1.5                 6.8
1  Mercury          0   0.4                 4.9

In [36]:
# TO DO: re-sort the DataFrame by the number of satellite in the decreasing order.
df.sort_values(by = 'Satellite', ascending = False)

Name  Satellite    AU  Diameter (in 1Kkm)
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2
4     Mars          2   1.5                 6.8
3    Earth          1   1.0                12.7
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1

Before 2006, Pluto was classified as a planet of the solar system. Let's add Pluto into our DataFrame.

In [37]:
df.loc[9] = ["Pluto", 0, 39.5, 2.38]
df

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                4.90
2    Venus          0   0.7               12.10
3    Earth          1   1.0               12.70
4     Mars          2   1.5                6.80
5  Jupiter         79   5.2              139.80
6   Saturn         60   9.5              116.50
7   Uranus         27  19.2               50.70
8  Neptune         14  30.1               49.20
9    Pluto          0  39.5                2.38

Let's reclassify pluto as a dwarf planet.

In [43]:
# Beware again! Remember that the 0th index of df.index is 1 ?
# To drop pluto, you either do df.drop(index=idx) or df.drop(df.index[idx])
# df.drop(9)
df.drop(df.index[8])

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1
3    Earth          1   1.0                12.7
4     Mars          2   1.5                 6.8
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

### 2.2 Read Data via Pandas
Pandas supports loading, reading, and writing data from various file format, including CSV, JSON and SQL, by converting it to a DataFrame. 
1. `pd.read_csv()` : Read CSV files
2. `pd.read_json()` : Read JSON files
3. `pd.read_sql_query()` : Read SQL files

In [61]:
# download data at https://support.spatialkey.com/spatialkey-sample-csv-data/
movie = pd.read_csv("./data/IMDB-Movie-Data.csv", index_col = "Title")   # ---> option 1
# movie = pd.read_csv("./data/IMDB-Movie-Data.csv")    #---> option 2
print(type(movie))
movie

<class 'pandas.core.frame.DataFrame'>


Rank                       Genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
Mindhorn                                    8                      Comedy   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Hounds of Love                             23          Crime,Drama,Horror   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Paris pieds nus                            26                      Comedy   
Bahubali: The Beginning                    27      Action,Adventure,Drama   
Dead Awake                                 28             Horror,Thriller   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
...                                       ...                         ...   
Texas Chainsaw 3D                         971             Horror,Thriller   
Disturbia                                 972      Drama,Mystery,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
Dark Places                               977      Drama,Mystery,Thriller   
Amateur Night                             978                      Comedy   
It's Only the End of the World            979                       Drama   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                       985            Adventure,Family   
Your Highness                             986    Adventure,Comedy,Fantasy   
Final Destination 5                       987             Horror,Thriller   
Endless Love                              988               Drama,Romance   
Martyrs                                   989                      Horror   
Selma                                

In [62]:
# To Do: Extract third row
# To Do: Uncomment the option 2 from the upper cell see which one works and which one does not
movie.iloc[2]
# movie.loc["Split"]
# movie.loc[2]

Rank                                                                  3
Genre                                                   Horror,Thriller
Description           Three girls are kidnapped by a man with a diag...
Director                                             M. Night Shyamalan
Actors                James McAvoy, Anya Taylor-Joy, Haley Lu Richar...
Year                                                               2016
Runtime (Minutes)                                                   117
Rating                                                              7.3
Votes                                                            157606
Revenue (Millions)                                               138.12
Metascore                                                            62
Name: Split, dtype: object

In [63]:
# To Do: Sort the table by Ratings, in the descending order
# Do you agree with the rankings? :)
movie.sort_values(by = "Rating", ascending = False)

Rank                       Genre  \
Title                                                                    
The Dark Knight                         55          Action,Crime,Drama   
Inception                               81     Action,Adventure,Sci-Fi   
Dangal                                 118      Action,Biography,Drama   
Interstellar                            37      Adventure,Drama,Sci-Fi   
Kimi no na wa                           97     Animation,Drama,Fantasy   
The Intouchables                       250      Biography,Comedy,Drama   
Whiplash                               134                 Drama,Music   
The Prestige                            65        Drama,Mystery,Sci-Fi   
The Departed                           100        Crime,Drama,Thriller   
Taare Zameen Par                       992          Drama,Family,Music   
The Dark Knight Rises                  125             Action,Thriller   
The Lives of Others                    477              Drama,Thriller   
WALL·E                                 635  Animation,Adventure,Family   
3 Idiots                               431                Comedy,Drama   
Koe no katachi                         862     Animation,Drama,Romance   
Django Unchained                       145               Drama,Western   
La La Land                               7          Comedy,Drama,Music   
Jagten                                 456                       Drama   
Up                                     500  Animation,Adventure,Comedy   
Inglourious Basterds                    78         Adventure,Drama,War   
Toy Story 3                            689  Animation,Adventure,Comedy   
Paint It Black                         479                       Drama   
Bahubali: The Beginning                 27      Action,Adventure,Drama   
PK                                     766        Comedy,Drama,Romance   
The Wolf of Wall Street                 83      Biography,Comedy,Crime   
Inside Out                             242  Animation,Adventure,Comedy   
Gran Torino                            646                       Drama   
Incendies                              714           Drama,Mystery,War   
El secreto de sus ojos                 743       Drama,Mystery,Romance   
Hacksaw Ridge                           17     Biography,Drama,History   
...                                    ...                         ...   
Final Girl                             778             Action,Thriller   
Dead Awake                              28             Horror,Thriller   
Viking                                 523        Action,Drama,History   
Zoolander 2                            432                      Comedy   
The Boy Next Door                      556            Mystery,Thriller   
Wolves at the Door                     820             Horror,Thriller   
Max Steel                              788     Action,Adventure,Family   
Only for One Night                     747                    Thriller   
Shut In                                827       Drama,Horror,Thriller   
The Assignment                         772       Action,Crime,Thriller   
The Human Centipede (First Sequence)   937                      Horror   
Sex and the City 2                     953        Comedy,Drama,Romance   
Yoga Hosers                            801       Comedy,Fantasy,Horror   
Movie 43                               509              Comedy,Romance   
Fantastic Four                         553     Action,Adventure,Sci-Fi   
Exposed                                839         Crime,Drama,Mystery   
The Last Airbender                     582     Action,Adventure,Family   
Fifty Shades of Grey                    64      Drama,Romance,Thriller   
2307: Winter's Dream                   617                      Sci-Fi   
The Disappointments Room               238       Drama,Horror,Thriller   
The Black Room                         402                      Horror   
Birth of the Dragon                    526

In [68]:
# To Do: Sort the table again by 'Revenue(Millions)', in ascending order and print the first 3 rows out
movie.sort_values(by = "Revenue (Millions)", ascending = True).head(3)

Rank                  Genre  \
Title                                           
A Kind of Murder   232   Crime,Drama,Thriller   
Into the Forest    962  Drama,Sci-Fi,Thriller   
Love, Rosie        678         Comedy,Romance   

                                                        Description  \
Title                                                                 
A Kind of Murder  In 1960s New York, Walter Stackhouse is a succ...   
Into the Forest   After a massive power outage, two sisters lear...   
Love, Rosie       Rosie and Alex have been best friends since th...   

                          Director  \
Title                                
A Kind of Murder      Andy Goddard   
Into the Forest    Patricia Rozema   
Love, Rosie       Christian Ditter   

                                                             Actors  Year  \
Title                                                                       
A Kind of Murder  Patrick Wilson, Jessica Biel, Haley Bennett, V...  2016   
Into the Forest   Ellen Page, Evan Rachel Wood, Max Minghella,Ca...  2015   
Love, Rosie       Lily Collins, Sam Claflin, Christian Cooke, Ja...  2014   

                  Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                    
A Kind of Murder                 95     5.2   3305                0.00   
Into the Forest                 101     5.9  10220                0.01   
Love, Rosie                     102     7.2  80415                0.01   

                  Metascore  
Title                        
A Kind of Murder       50.0  
Into the Forest        59.0  
Love, Rosie            44.0

In [76]:
# The value_counts() function is used to get a Series containing counts of unique values
movie['Genre'].value_counts().head()

Action,Adventure,Sci-Fi    50
Drama                      48
Comedy,Drama,Romance       35
Comedy                     32
Drama,Romance              31
Name: Genre, dtype: int64

In [82]:
# This is called a "Masking Operation"
movie[movie['Runtime (Minutes)'] >= 170].sort_values(by="Rating", ascending=False)

Rank                   Genre  \
Title                                                   
3 Idiots                  431            Comedy,Drama   
The Wolf of Wall Street    83  Biography,Comedy,Crime   
The Hateful Eight          89     Crime,Drama,Mystery   
La vie d'Adèle            312           Drama,Romance   
Grindhouse                829  Action,Horror,Thriller   
Cloud Atlas               268            Drama,Sci-Fi   
Inland Empire             966  Drama,Mystery,Thriller   

                                                               Description  \
Title                                                                        
3 Idiots                 Two friends are searching for their long lost ...   
The Wolf of Wall Street  Based on the true story of Jordan Belfort, fro...   
The Hateful Eight        In the dead of a Wyoming winter, a bounty hunt...   
La vie d'Adèle           Adèle's life is changed when she meets Emma, a...   
Grindhouse               Quentin Tarantino and Robert Rodriguez's homag...   
Cloud Atlas              An exploration of how the actions of individua...   
Inland Empire            As an actress starts to adopt the persona of h...   

                                    Director  \
Title                                          
3 Idiots                     Rajkumar Hirani   
The Wolf of Wall Street      Martin Scorsese   
The Hateful Eight          Quentin Tarantino   
La vie d'Adèle           Abdellatif Kechiche   
Grindhouse                  Robert Rodriguez   
Cloud Atlas                       Tom Tykwer   
Inland Empire                    David Lynch   

                                                                    Actors  \
Title                                                                        
3 Idiots                   Aamir Khan, Madhavan, Mona Singh, Sharman Joshi   
The Wolf of Wall Street  Leonardo DiCaprio, Jonah Hill, Margot Robbie,M...   
The Hateful Eight        Samuel L. Jackson, Kurt Russell, Jennifer Jaso...   
La vie d'Adèle           Léa Seydoux, Adèle Exarchopoulos, Salim Kechio...   
Grindhouse               Kurt Russell, Rose McGowan, Danny Trejo, Zoë Bell   
Cloud Atlas               Tom Hanks, Halle Berry, Hugh Grant, Hugo Weaving   
Inland Empire            Laura Dern, Jeremy Irons, Justin Theroux, Karo...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
3 Idiots                 2009                170     8.4  238789   
The Wolf of Wall Street  2013                180     8.2  865134   
The Hateful Eight        2015                187     7.8  341170   
La vie d'Adèle           2013                180     7.8  103150   
Grindhouse               2007                191     7.6  160350   
Cloud Atlas              2012                172     7.5  298651   
Inland Empire            2006                180     7.0   44227   

                         Revenue (Millions)  Metascore  
Title                                                   
3 Idiots                               6.52       67.0  
The Wolf of Wall Street              116.87       75.0  
The Hateful Eight                     54.12       68.0  
La vie d'Adèle                         2.20       88.0  
Grindhouse                            25.03        NaN  
Cloud Atlas                           27.10       55.0  
Inland Empire                           NaN        NaN

In [84]:
# To Do: Extract the movies whose 'Metascore' is bigger than 95, and list it from the recent to the least recent
movie[movie['Metascore'] >= 95].sort_values(by="Year", ascending=False)

Rank                    Genre  \
Title                                                  
Manchester by the Sea    22                    Drama   
Moonlight                42                    Drama   
Carol                   502            Drama,Romance   
Boyhood                 657                    Drama   
12 Years a Slave        112  Biography,Drama,History   
Gravity                 510    Drama,Sci-Fi,Thriller   
Zero Dark Thirty        407   Drama,History,Thriller   
The Social Network      325          Biography,Drama   
Ratatouille             490  Animation,Comedy,Family   
Pan's Labyrinth         231        Drama,Fantasy,War   

                                                             Description  \
Title                                                                      
Manchester by the Sea  A depressed uncle is asked to take care of his...   
Moonlight              A chronicle of the childhood, adolescence and ...   
Carol                  An aspiring photographer develops an intimate ...   
Boyhood                The life of Mason, from early childhood to his...   
12 Years a Slave       In the antebellum United States, Solomon North...   
Gravity                Two astronauts work together to survive after ...   
Zero Dark Thirty       A chronicle of the decade-long hunt for al-Qae...   
The Social Network     Harvard student Mark Zuckerberg creates the so...   
Ratatouille            A rat who can cook makes an unusual alliance w...   
Pan's Labyrinth        In the falangist Spain of 1944, the bookish yo...   

                                 Director  \
Title                                       
Manchester by the Sea    Kenneth Lonergan   
Moonlight                   Barry Jenkins   
Carol                         Todd Haynes   
Boyhood                 Richard Linklater   
12 Years a Slave            Steve McQueen   
Gravity                    Alfonso Cuarón   
Zero Dark Thirty          Kathryn Bigelow   
The Social Network          David Fincher   
Ratatouille                     Brad Bird   
Pan's Labyrinth        Guillermo del Toro   

                                                                  Actors  \
Title                                                                      
Manchester by the Sea  Casey Affleck, Michelle Williams, Kyle Chandle...   
Moonlight              Mahershala Ali, Shariff Earp, Duan Sanderson, ...   
Carol                  Cate Blanchett, Rooney Mara, Sarah Paulson, Ky...   
Boyhood                Ellar Coltrane, Patricia Arquette, Ethan Hawke...   
12 Years a Slave       Chiwetel Ejiofor, Michael Kenneth Williams, Mi...   
Gravity                Sandra Bullock, George Clooney, Ed Harris, Ort...   
Zero Dark Thirty       Jessica Chastain, Joel Edgerton, Chris Pratt, ...   
The Social Network     Jesse Eisenberg, Andrew Garfield, Justin Timbe...   
Ratatouille             Brad Garrett, Lou Romano, Patton Oswalt,Ian Holm   
Pan's Labyrinth        Ivana Baquero, Ariadna Gil, Sergi López,Maribe...   

                       Year  Runtime (Minutes)  Rating   Votes  \
Title                                                            
Manchester by the Sea  2016                137     7.9  134213   
Moonlight              2016                111     7.5  135095   
Carol                  2015                118     7.2   77995   
Boyhood                2014                165     7.9  286722   
12 Years a Slave       2013                134     8.1  486338   
Gravity                2013                 91     7.8  622089   
Zero Dark Thirty       2012                157     7.4  226661   
The Social Network     2010                120     7.7  510100   
Ratatouille            2007                111     8.0  504039   
Pan's Labyrinth        2006                118     8.2  498879   

                       Revenue (Millions)  Metascore  
Title                                                 
Manchester by the Sea               47.70       96.0  
Moonlight                

In [85]:
# To Do: Extract movies whose directed by one of UCL Alumni
movie[(movie['Director'] == 'Christopher Nolan')]

Rank                    Genre  \
Title                                                  
Interstellar             37   Adventure,Drama,Sci-Fi   
The Dark Knight          55       Action,Crime,Drama   
The Prestige             65     Drama,Mystery,Sci-Fi   
Inception                81  Action,Adventure,Sci-Fi   
The Dark Knight Rises   125          Action,Thriller   

                                                             Description  \
Title                                                                      
Interstellar           A team of explorers travel through a wormhole ...   
The Dark Knight        When the menace known as the Joker wreaks havo...   
The Prestige           Two stage magicians engage in competitive one-...   
Inception              A thief, who steals corporate secrets through ...   
The Dark Knight Rises  Eight years after the Joker's reign of anarchy...   

                                Director  \
Title                                      
Interstellar           Christopher Nolan   
The Dark Knight        Christopher Nolan   
The Prestige           Christopher Nolan   
Inception              Christopher Nolan   
The Dark Knight Rises  Christopher Nolan   

                                                                  Actors  \
Title                                                                      
Interstellar           Matthew McConaughey, Anne Hathaway, Jessica Ch...   
The Dark Knight        Christian Bale, Heath Ledger, Aaron Eckhart,Mi...   
The Prestige           Christian Bale, Hugh Jackman, Scarlett Johanss...   
Inception              Leonardo DiCaprio, Joseph Gordon-Levitt, Ellen...   
The Dark Knight Rises  Christian Bale, Tom Hardy, Anne Hathaway,Gary ...   

                       Year  Runtime (Minutes)  Rating    Votes  \
Title                                                             
Interstellar           2014                169     8.6  1047747   
The Dark Knight        2008                152     9.0  1791916   
The Prestige           2006                130     8.5   913152   
Inception              2010                148     8.8  1583625   
The Dark Knight Rises  2012                164     8.5  1222645   

                       Revenue (Millions)  Metascore  
Title                                                 
Interstellar                       187.99       74.0  
The Dark Knight                    533.32       82.0  
The Prestige                        53.08       66.0  
Inception                          292.57       74.0  
The Dark Knight Rises              448.13       78.0

#### 2.2.1 Pandas Exercise

To Do: Extract the movie list that meets below requirements:
- 1. Released in and after 2010 (key = 'Year')
- 2. Runtime is shorter than 150 minutes (key = 'Runtime (Minutes)')
- 3. Rating is above 8.0 (key = 'Rating') 
Print out only the first 3 movies.

In [114]:
movie[
    (movie['Year'] >= 2010) & 
    (movie['Runtime (Minutes)'] <= 150) & 
    (movie['Rating'] >= 8.0)
]

Rank  \
Title                                                
Guardians of the Galaxy                          1   
La La Land                                       7   
Hacksaw Ridge                                   17   
Lion                                            19   
Arrival                                         20   
Deadpool                                        34   
Star Wars: Episode VII - The Force Awakens      51   
Mad Max: Fury Road                              68   
Zootopia                                        75   
The Avengers                                    77   
Inception                                       81   
Gone Girl                                       84   
The Help                                        93   
Kimi no na wa                                   97   
The Martian                                    103   
12 Years a Slave                               112   
Harry Potter and the Deathly Hallows: Part 2   115   
Whiplash                                       134   
Shutter Island                                 139   
Room                                           144   
Ah-ga-ssi                                      146   
Twin Peaks: The Missing Pieces                 155   
X-Men: Days of Future Past                     163   
Her                                            174   
Forushande                                     185   
Spotlight                                      193   
Warrior                                        195   
The Imitation Game                             199   
The Perks of Being a Wallflower                236   
The Grand Budapest Hotel                       239   
Inside Out                                     242   
The Intouchables                               250   
Black Swan                                     262   
Sing Street                                    274   
Dallas Buyers Club                             333   
Rush                                           335   
Jagten                                         456   
Paint It Black                                 479   
Mommy                                          609   
Relatos salvajes                               642   
The King's Speech                              686   
Toy Story 3                                    689   
Incendies                                      714   
How to Train Your Dragon                       773   
Koe no katachi                                 862   

                                                                   Genre  \
Title                                                                      
Guardians of the Galaxy                          Action,Adventure,Sci-Fi   
La La Land                                            Comedy,Drama,Music   
Hacksaw Ridge                                    Biography,Drama,History   
Lion                                                     Biography,Drama   
Arrival                                             Drama,Mystery,Sci-Fi   
Deadpool                                         Action,Adventure,Comedy   
Star Wars: Episode VII - The Force Awakens      Action,Adventure,Fantasy   
Mad Max: Fury Road                               Action,Adventure,Sci-Fi   
Zootopia                                      Animation,Adventure,Comedy   
The Avengers                                               Action,Sci-Fi   
Inception                                        Action,Adventure,Sci-Fi   
Gone Girl                                            Crime,Drama,Mystery   
The Help                                                           Drama   
Kimi no na wa                                    Animation,Drama,Fantasy   
The Martian                                       Adventure,Drama,Sci-Fi   
12 Years a Slave                                 Biography,Drama,History   
Harry Potter and the Deathly Hallows: Part 2     Adventure,Drama,Fantasy   
Whiplash                                                     Dr

### 2.3 How to deal with Missing Data
To represent missing data, pandas use np.nan. Data scientists and machine learning engineers sometimes just remove missing data. However, it heavily depends on which data are missing, how big the missing data are and so on. You can fill the missing part with 0, with the mean value of the column or with mean value of only 10 nearest value in the column. It is important for you to choose the way how you are going to deal with missing data.
- `isnull()`: returns True or False, depending on the cell's null status. 
- `sum()`: This is a trick when you count the number of True's. Once the Dataframe is filtered through isnull() function, sum of all True's in a column gives you how many fields have missing data in them.
- `dropna()`: deletes any row that contains any single null value.
- `fillna(value)`: Fill missing value with the given values.

In [115]:
movie.isnull()

Rank  Genre  Description  Director  \
Title                                                                          
Guardians of the Galaxy                  False  False        False     False   
Prometheus                               False  False        False     False   
Split                                    False  False        False     False   
Sing                                     False  False        False     False   
Suicide Squad                            False  False        False     False   
The Great Wall                           False  False        False     False   
La La Land                               False  False        False     False   
Mindhorn                                 False  False        False     False   
The Lost City of Z                       False  False        False     False   
Passengers                               False  False        False     False   
Fantastic Beasts and Where to Find Them  False  False        False     False   
Hidden Figures                           False  False        False     False   
Rogue One                                False  False        False     False   
Moana                                    False  False        False     False   
Colossal                                 False  False        False     False   
The Secret Life of Pets                  False  False        False     False   
Hacksaw Ridge                            False  False        False     False   
Jason Bourne                             False  False        False     False   
Lion                                     False  False        False     False   
Arrival                                  False  False        False     False   
Gold                                     False  False        False     False   
Manchester by the Sea                    False  False        False     False   
Hounds of Love                           False  False        False     False   
Trolls                                   False  False        False     False   
Independence Day: Resurgence             False  False        False     False   
Paris pieds nus                          False  False        False     False   
Bahubali: The Beginning                  False  False        False     False   
Dead Awake                               False  False        False     False   
Bad Moms                                 False  False        False     False   
Assassin's Creed                         False  False        False     False   
...                                        ...    ...          ...       ...   
Texas Chainsaw 3D                        False  False        False     False   
Disturbia                                False  False        False     False   
Rock of Ages                             False  False        False     False   
Scream 4                                 False  False        False     False   
Queen of Katwe                           False  False        False     False   
My Big Fat Greek Wedding 2               False  False        False     False   
Dark Places                              False  False        False     False   
Amateur Night                            False  False        False     False   
It's Only the End of the World           False  False        False     False   
The Skin I Live In                       False  False        False     False   
Miracles from Heaven                     False  False        False     False   
Annie                                    False  False        False     False   
Across the Universe                      False  False        False     False   
Let's Be Cops                            False  False        False     False   
Max                                      False  False        False     False   
Your Highness                            False  False        False     False   
Final Destination 5                      False  False        False     False   
Endless Love                             F

In [116]:
movie.isnull().sum()

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

In [117]:
movie.shape

(1000, 11)

In [120]:
# Take a look at "Take Me Home Tonight" and "Search Party"
movie.fillna(value = 0)

Rank                       Genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
Mindhorn                                    8                      Comedy   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Hounds of Love                             23          Crime,Drama,Horror   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Paris pieds nus                            26                      Comedy   
Bahubali: The Beginning                    27      Action,Adventure,Drama   
Dead Awake                                 28             Horror,Thriller   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
...                                       ...                         ...   
Texas Chainsaw 3D                         971             Horror,Thriller   
Disturbia                                 972      Drama,Mystery,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
Dark Places                               977      Drama,Mystery,Thriller   
Amateur Night                             978                      Comedy   
It's Only the End of the World            979                       Drama   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                       985            Adventure,Family   
Your Highness                             986    Adventure,Comedy,Fantasy   
Final Destination 5                       987             Horror,Thriller   
Endless Love                              988               Drama,Romance   
Martyrs                                   989                      Horror   
Selma                                

In [121]:
movie.dropna(inplace = True)

In [122]:
movie.shape

(838, 11)

After dropping the rows that contain missing data, the shape of the dataFrame has changed, from (1000, 11) to (838, 11)

### 2.4 Merging Data
Some of you who know SQL might have felt that pandas is quite similar to query language.
What is the most popular thing that you do in most of the relational database query language?  
Yes! (terminologies alert!) Inner JOIN, Outer JOIN, Left JOIN, Right JOIN, Full JOIN...
- `concat()` : Concatenation. Used to merge two or more pandas object.
- `merge()` : It is merged in the way of SQL. 

In [123]:
df1 = pd.DataFrame(np.random.randn(10, 2))
df1

0         1
0 -1.292454  0.994892
1  0.407480 -0.693513
2  1.548593  0.114417
3 -0.362648  0.784441
4 -0.272858  1.829562
5  1.569398 -1.718500
6 -0.357931 -0.056753
7  0.977494 -0.576888
8 -0.510238  0.989765
9  0.993332  2.750075

In [124]:
df2 = pd.DataFrame(np.random.randn(10, 3))
df2

0         1         2
0  0.412625  0.605960 -1.821624
1 -0.186185  0.729974 -0.510493
2 -1.883591  0.250437  1.790063
3 -0.367597  1.338071 -1.112454
4  0.928728  0.839511 -2.002918
5  0.806188  3.427185  1.479313
6  1.392479 -1.748136 -0.655155
7  0.107495  0.419968 -0.155408
8 -0.281937 -1.269672 -1.369113
9 -0.199018 -0.331693  0.584287

In [125]:
pd.concat([df1, df2])

0         1         2
0 -1.292454  0.994892       NaN
1  0.407480 -0.693513       NaN
2  1.548593  0.114417       NaN
3 -0.362648  0.784441       NaN
4 -0.272858  1.829562       NaN
5  1.569398 -1.718500       NaN
6 -0.357931 -0.056753       NaN
7  0.977494 -0.576888       NaN
8 -0.510238  0.989765       NaN
9  0.993332  2.750075       NaN
0  0.412625  0.605960 -1.821624
1 -0.186185  0.729974 -0.510493
2 -1.883591  0.250437  1.790063
3 -0.367597  1.338071 -1.112454
4  0.928728  0.839511 -2.002918
5  0.806188  3.427185  1.479313
6  1.392479 -1.748136 -0.655155
7  0.107495  0.419968 -0.155408
8 -0.281937 -1.269672 -1.369113
9 -0.199018 -0.331693  0.584287

In [128]:
pd.concat([df1, df2], axis = 1)

0         1         0         1         2
0 -1.292454  0.994892  0.412625  0.605960 -1.821624
1  0.407480 -0.693513 -0.186185  0.729974 -0.510493
2  1.548593  0.114417 -1.883591  0.250437  1.790063
3 -0.362648  0.784441 -0.367597  1.338071 -1.112454
4 -0.272858  1.829562  0.928728  0.839511 -2.002918
5  1.569398 -1.718500  0.806188  3.427185  1.479313
6 -0.357931 -0.056753  1.392479 -1.748136 -0.655155
7  0.977494 -0.576888  0.107495  0.419968 -0.155408
8 -0.510238  0.989765 -0.281937 -1.269672 -1.369113
9  0.993332  2.750075 -0.199018 -0.331693  0.584287

In [129]:
demis = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Demis' : [75, 97, 64, 81]}
)
demis

Modules  Demis
0   Bioinformatics     75
1  Robotic Systems     97
2         Security     64
3        Compilers     81

In [130]:
sedol = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Sedol' : [63, 78, 84, 95]})
sedol

Modules  Sedol
0   Bioinformatics     63
1  Robotic Systems     78
2         Security     84
3        Compilers     95

In [131]:
pd.merge(demis, sedol, on = 'Modules')

Modules  Demis  Sedol
0   Bioinformatics     75     63
1  Robotic Systems     97     78
2         Security     64     84
3        Compilers     81     95

In [132]:
#To Do: Define your own dataframe and use functions introducesd above to concatenate them.

### What to do next?
Below websites would be helpful for your further study on pandas library:
- [Pandas official website](https://pandas.pydata.org)
- [10 minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html)
- [Data Wrangling with Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)